Fix #1489: responsive conversation skin

maitbayev opened the pull request:
Hi @seanlip,
This is first try on #1489. The pull request is intended to show progress on the issue and get some suggestions, rather than review. I am ready for comments like "this is wrong, please close it". 
This PR doesn't handle all cases of amazing spec doc prepared by @amitdeutsch even when viewport is >= 960.
All supplementary interactions should be re-designed to be fixed, they can't shrink now.
In addition, I couldn't come up with any solution of "shrink tutor card first then supplementary".
One idea, to use "flex-shrink" css attribute to put "flex-shrink" of tutor card much higher than supplementary's, but I don't like very much it isn't deterministic solution.

Also, Github shows that I have changed a lot, in fact I didn't change a lot.

What do you think? What do you suggest?

seanlip commented:
Hi @maitbayev -- I took a look, but I can't give you much feedback, since I'm a little confused about the scope of the PR. For example, it seems weird to have both blue buttons at the bottom for large screens, since they don't do anything. But you were talking about 960 >= viewport, which also looks a bit off (I thought that, in that case, one of the cards would go above the other).

So, could you please clarify what this PR is attempting to do? Then I can comment better.

About the shrinking, I agree that that is tricky. Maybe we need to use JavaScript? I don't think there's a native solution in CSS.

Also, I would suggest hardcoding one of the interactions (Pencil Code, say) to be a fixed 1000px width, and one of the interactions to be a fixed 560px width (just for the purposes of this PR). That would make it easy to quickly test the PR using those two, and see if things work properly in both cases -- you want a fast feedback loop for this sort of thing. We can always revert those changes later before the PR is merged.

Please let me know if you have any other questions, or if I haven't answered yours adequately!

maitbayev commented:
Hi @seanlip,
Ah, actually for viewport >= 960
I got confused with signs

seanlip commented:
Ah, ok, thanks! My other comments still apply, I think.

maitbayev commented:
I will double check later when I will have access to a computer.
But what do you think about using flex ?

maitbayev commented:
Hey @seanlip,
Take a look on this commit,
seems like it works.

seanlip commented:
Hm, I would prefer something more deterministic, so, given what you mentioned, I'm not sure we can do what we want to do with flex. JS may be better.

maitbayev commented:
@seanlip, have you tried on practice ?

seanlip commented:
Oh sorry, my reply was to the comment before your last reply -- I didn't see your new commit. Let me take a look!

maitbayev commented:
@seanlip, but it is still the same approach. Having said that, your response also applies to a new commit:

>  I would prefer something more deterministic, so, given what you mentioned, I'm not sure we can do what we want to do with flex

seanlip commented:
I just tested, and it looks good, UX-wise. The 'O' avatar tends to get a bit cut off on the left at the smaller sizes but it's fine to tweak that in a separate pass.

So, I was mistaken, sorry -- I based on my response on your mention of non-determinism, and my previous experiments with flex. But the approach you've taken seems fine to me, since what matters is how the user experiences it.

maitbayev commented:
I noticed 'O' avatar, but the design specs doesn't include it. What do you suggest as a fixing for 'O' avatar ?

seanlip commented:
Also, while this looks fine for width >= 960px, it breaks things for width < 960px -- so, do you think we could handle the latter case in this same PR too, before merging into develop? Otherwise, if we merge this as-is, it'll break the experience on smaller screens.

seanlip commented:
Not sure yet, tbh. But, in the grand scheme of things, it feels minor, so I figured I'd check with @amitdeutsch later about it.

Though, I guess, no time like the present :-) So -- @amitdeutsch, a question for you: with the new responsive design for the learner view, the avatar gets cut off around the 960px boundary for larger screens. Should we hide it once it starts getting cut off, or push it into the card, or increase the margins, or something else? (See screenshot below.)

![responsive](https://cloud.githubusercontent.com/assets/10575562/14770507/2b07773c-0a28-11e6-8af3-474609e7c1b2.png)

Thanks!

maitbayev commented:
@seanlip,
FYI, this PR also includes targeted exploration which you can load in the admin page.

for viewport < 960 I will try, but a lot of css which are similar to each other and difficult to hold in my memory at the same time.

seanlip commented:
Ah yep, I saw that, and it was really helpful! I included a screenshot of it in my recent comment.

Btw, feel free to refactor the existing CSS as needed, if it makes things easier -- I do realize that there's a fair number of rules and it's not conceptually scaffolded that well at the moment.

maitbayev commented:
I have a little cleaned css, on the way some questions raised, can you please confirm my assumptions:

1. conversation-skin-tutor-card-feedback-toggle-* : we don't need anymore, as `grep "conversation-skin-tutor-card-feedback-toggle" ` showed me only lines inside the <style>. So if you know where we used it, can I delete it safely ?
2. conversation-skin-card-switcher-*: deleted as we won't have switcher. 
3.  div.conversation-skin-supplemental-card : deleted it as it doesn't look relevant. What does this css selector mean? 
4. Deleted `@media` : they are going to change anyway.

maitbayev commented:
I have a little bit cleaned css, so some questions araised, can you

maitbayev commented:
Also,
I found out that old html wasn't valid ( i.e two extra closed div's). 
So, is there any tool that can test for validity for me ? I tried several chrome extensions but seems like they work poor with angular. I am interested in opening and closing tags.

seanlip commented:
1. Agreed -- I don't see where it's being used either.
2. Sgtm.
3. The supplemental card refers to the card on the right when there are two cards side-by-side. It means a div that also has a "conversation-skin-supplemental-card" class. I'm not entirely certain why the "div" is there -- perhaps it was to force the rule to come higher in the CSS priority list (the more specific the rule, the higher its priority).
4. OK.

Re testing validity -- I don't know, sorry. There's stuff like [this](http://www.freeformatter.com/html-validator.html) but it kind of chokes on the ampersands that are used in angular expressions. Still, maybe one could just focus on the warnings which relate to open/close tags, and ignore the others.

maitbayev commented:
Hi @seanlip,
Please take a look on this commit.
This is first try on a narrow viewports.
It doesn't look great, so I need a lot of suggestions and improvements.

seanlip commented:
Hmm -- I loaded the Welcome exploration, and I actually think the tutor+supplemental interaction on narrow screens is not bad! However, the first three cards (where there's no supplemental card) simply don't show up on a narrow screen. Can you fix this?

Also, on the "pick Helsinki on a map" card, I get "Cannot read property 'top' of undefined" when I select a wrong answer.

(I'm not exactly sure if this is what you meant by suggestions/improvements, but in general I think it's better to get the basic functionality working first -- i.e. fixing the two points above. After that, we can work on the subtler positioning/colour details. Thanks!)

maitbayev commented:
Hi @seanlip,
They are fixed.
By suggestion I meant improvements in way of approach and code, as I am novice at html/css, as well as designs. For example, from view of design, I am not sure whether black mask behind the tutor card will work fine.
However, two issues you mentioned are not intended, so accordingly hopefully fixed.

seanlip commented:
Hi @maitbayev, thanks! This looks great!

Let's start with all the design/UX stuff, and then when that's fixed we can look into the tidying up the code a bit, as needed. Some things I noticed:

- In the narrow view, when there's a tutor card overlaid on top of the supplement, the instruction at the bottom should be a blue button (like the continue button) rather than instructions on a grey background with an arrow to the right. Otherwise, it's not clear to the user what they should click.
- The mask should not be black/opaque. It should probably be more like `#ccc` with opacity 0.7 or so.
- The mask should also match the background dimensions exactly -- in the screenshot below, the grey mask juts out from the background a little bit:
    ![1489-background-mask](https://cloud.githubusercontent.com/assets/10575562/14842240/2706b4e4-0bfa-11e6-8737-b123c032f5fb.png)
- It might be worth putting a bit of shadow below the tutor card when it's overlaid on the supplement, otherwise it's a bit hard to distinguish the two.
- It might also be worth putting a bit of shadow below the Oppia avatar when it's at the top right of the supplement, so that it looks obviously clickable.
- On wide screens, the animation that used to be present when going from the third state (one card on the screen) to the fourth card (two cards side-by-side) in the Welcome exploration seems to have now disappeared. It would be nice to keep it.
- On wide screens, the cards seem completely off-center:
    ![1489-help-weird](https://cloud.githubusercontent.com/assets/10575562/14842331/1a9fa46c-0bfb-11e6-8a83-00a3eb17ec6c.png)
- (Optional, but would be nice) Is it straightforward to try and do the minimize/restore animation of the tutor card on narrow screens? It would be nice to see how that looks, if possible.

Thanks!

maitbayev commented:
For myself (things to be done):

- [ ] In the narrow view, when there's a tutor card overlaid on top of the supplement, the instruction at the bottom should be a blue button (like the continue button) rather than instructions on a grey background with an arrow to the right. Otherwise, it's not clear to the user what they should click.

- [ ] The mask should not be black/opaque. It should probably be more like #ccc with opacity 0.7 or so.

- [ ] The mask should also match the background dimensions exactly -- in the screenshot below, the grey mask juts out from the background a little bit: 

- [ ] It might be worth putting a bit of shadow below the tutor card when it's overlaid on the supplement, otherwise it's a bit hard to distinguish the two.

- [ ] It might also be worth putting a bit of shadow below the Oppia avatar when it's at the top right of the supplement, so that it looks obviously clickable.

- [ ] On wide screens, the animation that used to be present when going from the third state (one card on the screen) to the fourth card (two cards side-by-side) in the Welcome exploration seems to have now disappeared. It would be nice to keep it.

- [ ] On wide screens, the cards seem completely off-center.

- [ ]  \(Optional, but would be nice\) Is it straightforward to try and do the minimize/restore animation of the tutor card on narrow screens? It would be nice to see how that looks, if possible.

maitbayev commented:
Hi @seanlip,
I think, I addressed all your comments except the last one (which I haven't started researching).
So, take an another look.

maitbayev commented:
**upd**: some basic animation is done.

seanlip commented:
Hi @maitbayev,

Thanks! I noticed a few more things:
* In the narrow view, with just a single card (no supplement), there shouldn't be a shadow:
  ![1754-1](https://cloud.githubusercontent.com/assets/10575562/14872957/1e59d220-0ca6-11e6-97a6-785fc92bc977.png)
* Play the welcome exploration all the way up to the map. Click on the tutor to view the map. Then click the back arrow to view previous cards. The Oppia avatar appears in an odd place:
  ![1754-2](https://cloud.githubusercontent.com/assets/10575562/14872974/44843fb2-0ca6-11e6-9454-d6ade3170008.png)
* When the user is on a supplement, and clicks the Oppia avatar, could you animate the tutor card out? (The animation you added is great, btw, and makes a big difference.)
* The white margins around the map look quite odd. Is it possible to just have the map part grayed out, with the surrounding area unaffected?
  ![1754-3](https://cloud.githubusercontent.com/assets/10575562/14873056/10327b42-0ca7-11e6-91be-137b7a7556f8.png)
* We should tidy up the supplemental interactions and make sure they all behave nicely. Would it be possible to do that? If it's simpler for them to be fixed-width for now, that's fine.
  * Also, please could you translate the design constraints into technical specifications for interaction developers so that their interactions work well in the learner view? (E.g., you must 'declare a "width: 600px;" on your outermost div' or something like that.)

seanlip commented:
@maitbayev -- also, could you make Pencil Code have width 1000px, and return CodeRepl to what it was before? Similarly, can you make the map 560px, and also make ImageClickInput responsive according to the size of the image? We ran into a little difficulty with testing because the CodeRepl image was unexpectedly wide and it looked a bit weird to the user.

Also, for wide (in fact, all) supplemental interactions, it would be better for the avatar be nearer the top right, rather than floating randomly somewhere in the middle:

![screenshot from 2016-04-27 21_38_59](https://cloud.githubusercontent.com/assets/10575562/14873156/22bd1ae6-0ca8-11e6-91e1-9c7f34084065.png)

Thanks!

/cc @jacobdavis11

jacobdavis11 commented:
If the screen is not tall enough then feedback messages on the map interaction are displayed below the bottom of it, even though they are floating over the map and could easily be moved up. See screenshot.
![screenshot](https://cloud.githubusercontent.com/assets/13726796/14873162/77131e06-0cc1-11e6-84d3-90078a4f3bd3.png)

maitbayev commented:
For my self:

- [ ] In the narrow view, with just a single card (no supplement), there shouldn't be a shadow

- [ ] Play the welcome exploration all the way up to the map. Click on the tutor to view the map. Then click the back arrow to view previous cards. The Oppia avatar appears in an odd place: 

- [ ] When the user is on a supplement, and clicks the Oppia avatar, could you animate the tutor card out? (The animation you added is great, btw, and makes a big difference.)

- [ ] The white margins around the map look quite odd. Is it possible to just have the map part grayed out, with the surrounding area unaffected? 

- [ ] Make Pencil Code have width 1000px, and return CodeRepl to what it was before

- [ ] Make the map 560px, and also make ImageClickInput responsive according to the size of the image

- [ ] Also, for wide (in fact, all) supplemental interactions, it would be better for the avatar be nearer the top right, rather than floating randomly somewhere in the middle.

- [ ] If the screen is not tall enough then feedback messages on the map interaction are displayed below the bottom of it, even though they are floating over the map and could easily be moved up

Wow, still a lot of things to do :-).

maitbayev commented:
Hi @seanlip,
I haven't finished all issues yet, but can you review what is done ?
Ah, I am starting to get frustrated with this PR :-). By trying to fix some parts, can add more issues. 
So hopefully, major issues fixed.

seanlip commented:
Ah, sorry to hear that! I think it's because this PR is essentially a whole bunch of different things wrapped into one, so that's why there's a lot of stuff to keep track of.

Having said that -- I played with the latest version and it is very nice! The only thing I noted is that we've somehow lost the animation in transitioning from the one-card to two-card view on large screens (e.g. the third to the fourth card in the Welcome exploration). However, I'd be fine with adding this in a separate PR, since I think the responsiveness is more important.

Will you be looking next into doing an audit pass on the individual interactions, so that they're reasonably presentable? I think that's the next step, and after that let's tidy up and merge.

Thanks!

maitbayev commented:
Hey @seanlip,
Animation should be back now.
Also almost all issues are fixed.
Can you take an another look ?

seanlip commented:
Hi @maitbayev, this looks great, thanks!

The only thing that's still problematic, I think, is the big white border around the map, and in the image in the sample exploration you created -- we should get rid of that.

Otherwise, I think this is almost ready for merge. I'll also take a look at the code now and make inline comments.

Thanks!

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -668,6 +688,35 @@ oppia.directive('conversationSkin', [function() {
+                 $scope._tutorCardDisplayedOnNarrow;
+        };
+
+        $scope.isSupplementalCardDisplayedOnNarrow = function() {
```
Maybe: isScreenNarrowAndShowingSupplementalCard

(otherwise it sounds like 'narrow' here is a pre-condition, rather than something we're checking for)

maitbayev replied:
done

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -668,6 +688,35 @@ oppia.directive('conversationSkin', [function() {
+          return $scope.windowWidth < TWO_CARD_THRESHOLD_PX;
+        };
+
+        $scope.isTutorCardDisplayedOnNarrow = function() {
```
isScreenNarrowAndDisplayingTutorCard

maitbayev replied:
Renamed to isScreenNarrowAndShowingTutorCard for consistency.

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -668,6 +688,35 @@ oppia.directive('conversationSkin', [function() {
           return $scope.windowWidth >= TWO_CARD_THRESHOLD_PX;
         };
 
+        $scope.isNarrowViewport = function() {
```
isViewportNarrow

maitbayev replied:
done

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -668,6 +688,35 @@ oppia.directive('conversationSkin', [function() {
+                 !$scope._tutorCardDisplayedOnNarrow;
+        };
+
+        $scope.displayTutorCardOnNarrow = function() {
```
displayTutorCardIfScreenIsNarrow

maitbayev replied:
renamed to displayTutorCardIfScreenIsNarrow

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -668,6 +688,35 @@ oppia.directive('conversationSkin', [function() {
+          }
+        };
+
+        $scope.displaySupplementaryCardOnNarrow = function() {
```
displaySupplementalCardIfScreenIsNarrow

maitbayev replied:
renamed to showSupplementaryCardOnNarrow

seanlip replied:
Note: 'supplemental', not 'supplementary'.

Also suggest 'on' --> 'if'. 'on' suggests an event handler.

maitbayev replied:
yeap, I did it right, just got confused here.

seanlip commented on core/templates/dev/head/player/conversation_skin_directive.html:
```diff
@@ -353,10 +322,32 @@
 
     .conversation-skin-cards-container {
       margin: 0 auto;
```
nit: in each css selector, alphabetize the attributes. (Here and below.)

maitbayev replied:
done

seanlip commented on extensions/interactions/InteractiveMap/InteractiveMap.html:
```diff
@@ -7,11 +7,27 @@
     }
+
+    #map_canvas {
+      height: 400px;
```
alphabetize

maitbayev replied:
done

seanlip commented on extensions/interactions/MusicNotesInput/static/css/musicNotesInput.css:
```diff
@@ -106,7 +106,7 @@
 .oppia-music-input-on-staff {
   position: absolute;
-  z-index: 1000;
+  z-index: 14;
```
Can you add a comment to warn developers why this shouldn't be changed in future (or what constraints they should be careful about here)? I assume you changed it for a reason.

maitbayev replied:
done

seanlip commented on extensions/interactions/PencilCodeEditor/PencilCodeEditor.html:
```diff
@@ -8,6 +8,14 @@
   </div>
+   <style>
+    .code-editor {
+      height: 500px;
```
alphabetize

seanlip commented:
Just took a code pass, and the code looks great in general -- I couldn't find much to comment on. Thanks, @maitbayev!

maitbayev commented:
Hi @seanlip,
For now, I have no idea how to fix white margins. Any thoughts?

seanlip commented:
I took a quick look, and it seems like it's an issue with the width of .supplemental-interaction-container (at least when I tried with the map). Why does it fill the whole screen? I don't think it should -- instead, it should take the relevant width with equal margins at the sides.

maitbayev commented:
@seanlip,
Tried different approaches, some of them works with map fine, but breaks other interactions.

maitbayev commented:
Hi @seanlip, 
Hooray, white margin should be fixed in this PR.
What we discussed via hangout didn't work.
But, it was part of this solution. 

Let me explain my approach:
There are two type of supplemental interactions: 
1. Width can be calculated independently  from container ('.conversation-skin-cards-container'), such as fixed width or width derived from content.
2. Width can not be calculated independently from container, such as elements who use (width=100%).

My solution is about using 100% of viewport in the second case.

Let me know if this approach is fine.

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -34,29 +34,105 @@ oppia.animation('.conversation-skin-responses-animate-slide', function() {
   };
 });
 
+var adjustTutorAvatar = function() {
```
'adjust' is a bit vague. Maybe repositionTutorAvatar?

maitbayev replied:
I found a better solution with css.  So this code is deleted.

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -281,57 +340,14 @@ oppia.directive('conversationSkin', [function() {
           $scope.clearHelpCard();
-
-          _recomputeAndResetPanels();
+          $scope._tutorCardDisplayedOnNarrow = true;
```
Drop leading underscore (we're moving away from that convention, see [style rules](https://github.com/oppia/oppia/blob/develop/CONTRIBUTING.md#style-rules)). Use var instead of $scope to indicate that this is local to this controller.

Also, please add initialization of this variable at the top (around line 285 or so) and a one-line comment explaining what it means, as precisely as possible. (Is it actually supposed to be tutorCardDisplayedIfNarrow? I'm not totally sure whether this means "display the tutor card once a screen-has-just-become-narrow event is received", or "display the tutor card whenever the screen is narrow".)

maitbayev replied:
done

seanlip commented on core/templates/dev/head/player/conversation_skin_directive.html:
```diff
@@ -29,18 +28,12 @@
+      <div ng-show='!isViewportNarrow() || isScreenNarrowAndShowingTutorCard()'
+           class="conversation-skin-main-tutor-card"
+           ng-class="{'animate-card-width': startCardChangeAnimation,
+                      'tutor-card-shadow': isViewportNarrow() && isCurrentSupplementalCardNonempty(),
```
I think, let's call this class 'has-shadow', since the CSS definition doesn't seem specific to the tutor card.

maitbayev replied:
done

seanlip commented on extensions/interactions/ImageClickInput/ImageClickInput.html:
```diff
@@ -1,14 +1,35 @@
          ng-click="onClickImage($event)"
          ng-mousemove="onMousemoveImage($event)"
-         style="cursor: pointer; display: block; margin-left: auto; margin-right: auto;">
+         style="">
```
This line is now unnecessary?

maitbayev replied:
done

seanlip commented on feconf.py:
```diff
@@ -339,6 +339,7 @@ def get_empty_ratings():
     u'14': 'about_oppia.yaml',
     u'15': 'fuzzy_exploration.yaml',
     u'16': 'all_interactions',
+    u'17': 'test_responsive_supplementary'
```
Can we now remove this exploration, and optionally merge it into the test_interactions exploration (number 10)? I think that's a better long-term place for the tests here, and it helps avoid a proliferation of demo explorations.

maitbayev replied:
done

seanlip commented on extensions/interactions/PencilCodeEditor/PencilCodeEditor.js:
```diff
@@ -33,8 +33,8 @@ oppia.directive('oppiaInteractivePencilCodeEditor', [
             $attrs.initialCodeWithValue);
-
-          var pce = new PencilCodeEmbed($element[0].children[0]);
+          var iframeDiv = document.getElementById('pencil-code-editor-iframe');
```
Mm, not so good -- this references the entire DOM, whereas the stuff that this extension does should really be encapsulated within the extension.

Can we instead give the div a class (not an id) and do something like $element[0].getSubElementByClassName? I.e. something like:

    http://stackoverflow.com/questions/6933763/how-do-i-select-item-with-class-within-a-div-in-jquery

maitbayev replied:
done

seanlip commented:
Hi @maitbayev -- yep, this approach is good. I just played with the UI, and I think it looks great!

I wrote a bunch of comments, but they should all be pretty easy to addressed. Once that's done, I think this can be merged in.

Thanks!

maitbayev commented:
Hey @seanlip,
Take an another look please.
I responded to your all comments.

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -190,6 +253,9 @@ oppia.directive('conversationSkin', [function() {
         var _answerIsBeingProcessed = false;
         var _nextFocusLabel = null;
+        // This variable is used only when viewport is narrow.
+        // Indicates whether tutor card is active or supplemental card
```
Replace this line with: "Indicates whether the tutor card is displayed."

(Note: the tutor card is, by definition, not the supplemental card.)

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -190,6 +253,9 @@ oppia.directive('conversationSkin', [function() {
         var _nextFocusLabel = null;
+        // This variable is used only when viewport is narrow.
+        // Indicates whether tutor card is active or supplemental card
+        var isTutorCardDisplayedIfNarrow = true;
```
nit: tutorCardIsDisplayedIfNarrow

(i.e. don't start with a verb)

seanlip commented:
Sorry -- I played through the whole demo exploration, and found one more thing. In develop, this is how the "help card" looks:

![help-card-develop](https://cloud.githubusercontent.com/assets/10575562/15132529/86e553ce-160f-11e6-9742-6eaf1219d2d7.png)

In this branch, this is how the "help card" looks:

![hello-oppia](https://cloud.githubusercontent.com/assets/10575562/15132518/68145bf2-160f-11e6-8fed-9ab0c6e9b438.png)

The 'develop' version is correct. Can the help card display on the bottom right of the supplemental interaction?

maitbayev commented:
Hi @seanlip,
It is also fixed. Take a look.

seanlip commented on core/templates/dev/head/player/conversation_skin_directive.html:
```diff
@@ -132,7 +132,7 @@
 
         <div ng-if="isScreenNarrowAndShowingTutorCard()" class="supplemental-mask"></div>
-        <div class="conversation-skin-help-card" ng-if="helpCardHtml">
+          <div class="conversation-skin-help-card" ng-if="helpCardHtml">
```
Oops -- indentation wrong

seanlip commented:
One small nit, otherwise LGTM! Feel free to merge once travis checks pass.

Thanks @maitbayev!

maitbayev commented:
@seanlip done.

seanlip commented:
Merging. Thanks @maitbayev!
